mixer/template/sample/createinstance_test.go

```diff
@@ -0,0 +1,985 @@
+		DoublePrimitive: "1.2",
+		Int64Primitive:  "54362",
+		StringPrimitive: `"mystring"`,
+		Int64Map:        map[string]string{"a": "1"},
+		TimeStamp:       "ats",
+		Duration:        "adr",
+		Res2: &sample_report.Res2InstanceParam{
+			Value:          "1",
+			Dimensions:     map[string]string{"s": "2"},
+			Int64Primitive: "54362",
+			DnsName:        `"myDNS"`,
+			Duration:       "adr",
+			EmailAddr:      `"myEMAIL"`,
+			IpAddr:         `ip("0.0.0.0")`,
+			TimeStamp:      "ats",
+			Uri:            `"myURI"`,
+		},
+		Res2Map: map[string]*sample_report.Res2InstanceParam{
+			"foo": {
+				Value:          "1",
+				Dimensions:     map[string]string{"s": "2"},
+				Int64Primitive: "54362",
+				DnsName:        `"myDNS"`,
+				Duration:       "adr",
+				EmailAddr:      `"myEMAIL"`,
+				IpAddr:         `ip("0.0.0.0")`,
+				TimeStamp:      "ats",
+				Uri:            `"myURI"`,
+			},
+		},
+	},
+}
+
+// given the default instance param and the default attributes, this is the expected instance. The name
+// "instance1" is hard-coded.
+var defaultReportInstance = &sample_report.Instance{
+	Name:            "instance1",
+	Value:           int64(1),
+	Dimensions:      map[string]interface{}{"s": int64(2)},
+	BoolPrimitive:   true,
+	DoublePrimitive: 1.2,
+	Int64Primitive:  54362,
+	StringPrimitive: "mystring",
+	Int64Map:        map[string]int64{"a": int64(1)},
+	TimeStamp:       time.Date(2017, time.January, 01, 0, 0, 0, 0, time.UTC),
+	Duration:        10 * time.Second,
+	Res1: &sample_report.Res1{
+		Value:           int64(1),
+		Dimensions:      map[string]interface{}{"s": int64(2)},
+		BoolPrimitive:   true,
+		DoublePrimitive: 1.2,
+		Int64Primitive:  54362,
+		StringPrimitive: "mystring",
+		Int64Map:        map[string]int64{"a": int64(1)},
+		TimeStamp:       time.Date(2017, time.January, 01, 0, 0, 0, 0, time.UTC),
+		Duration:        10 * time.Second,
+		Res2: &sample_report.Res2{
+			Value:          int64(1),
+			Dimensions:     map[string]interface{}{"s": int64(2)},
+			Int64Primitive: 54362,
+			DnsName:        adapter.DNSName("myDNS"),
+			Duration:       10 * time.Second,
+			EmailAddr:      adapter.EmailAddress("myEMAIL"),
+			IpAddr:         net.ParseIP("0.0.0.0"),
+			TimeStamp:      time.Date(2017, time.January, 01, 0, 0, 0, 0, time.UTC),
+			Uri:            adapter.URI("myURI"),
+		},
+		Res2Map: map[string]*sample_report.Res2{
+			"foo": {
+				Value:          int64(1),
+				Dimensions:     map[string]interface{}{"s": int64(2)},
+				Int64Primitive: 54362,
+				DnsName:        adapter.DNSName("myDNS"),
+				Duration:       10 * time.Second,
+				EmailAddr:      adapter.EmailAddress("myEMAIL"),
+				IpAddr:         net.ParseIP("0.0.0.0"),
+				TimeStamp:      time.Date(2017, time.January, 01, 0, 0, 0, 0, time.UTC),
+				Uri:            adapter.URI("myURI"),
+			},
+		},
+	},
+}
+
+func generateReportTests() []createInstanceTest {
+	var tests []createInstanceTest
+
+	r := defaultReportInstanceParam // make a copyAttrs
```

Comment:
suggestion: This is important step to make a copy and new tests added by others can accidentally skip it and mutate the original data. Can we really hard to debug since tests might run in parallel.

can we rename the field to something like `defaultReportInstanceParamDoNotUseDirectly` and have a function `getdefaultReportInstanceParam()` which returns a copy.

Reply:
As discussed offline, the mutation is during the test list construction, and does not happen during test execution. As such the failures due to copy problems are consistent.

I've considered writing such functions but decided against it eventually, as making deep-copies would require some meaty code.